### PR TITLE
Lockdep

### PR DESCRIPTION
**Pull request by jforissier**

### CODE REVIEW

**Comment by jforissier** [comment]
Here is the "lockdep" code I've been working on recently. To get a better idea of what it does, here is what I got by removing a random `mutex_unlock()` in the FS code ([here](https://github.com/OP-TEE/optee_os/blob/master/core/tee/tee_ree_fs.c#L960)). Interestingly, the problem is reported before the actual deadlock occurs (trying to lock the same lock a second time), because the missing unlock causes a violation in the locking order.

```
E/TC:? 0 Potential deadlock detected!
E/TC:? 0 When trying to acquire lock 0xe15d110
E/TC:? 0 -> Shortest cycle:
E/TC:? 0  Lock 0xe15d110
E/TC:? 0  Lock 0xe15d144
E/TC:? 0 -> Thread 0xe15da08 acquired lock 0xe15d144 at:
E/TC:? 0 Call stack:
E/TC:? 0  0xe129131 __lockdep_lock_acquire at optee_os/core/kernel/lockdep.c:309
E/TC:? 0  0xe11653d lockdep_lock_acquire at optee_os/core/include/kernel/lockdep.h:82
E/TC:? 0  0xe1166a9 mutex_lock_check at optee_os/core/arch/arm/kernel/mutex_lockdep.c:45
E/TC:? 0  0xe115e5b __mutex_lock at optee_os/core/arch/arm/kernel/mutex.c:31
E/TC:? 0  0xe116229 mutex_lock at optee_os/core/arch/arm/kernel/mutex.c:238
E/TC:? 0  0xe135959 ree_fs_open at optee_os/core/tee/tee_ree_fs.c:616
E/TC:? 0  0xe139cad tadb_open at optee_os/core/tee/tadb.c:214
E/TC:? 0  0xe139d67 tee_tadb_open at optee_os/core/tee/tadb.c:234
E/TC:? 0  0xe13a705 tee_tadb_ta_open at optee_os/core/tee/tadb.c:657
E/TC:? 0  0xe1107db secstor_ta_open at optee_os/core/arch/arm/kernel/secstor_ta.c:19
E/TC:? 0  0xe10f911 load_elf_from_store at optee_os/core/arch/arm/kernel/user_ta.c:726
E/TC:? 0  0xe10fc6d load_elf at optee_os/core/arch/arm/kernel/user_ta.c:843
E/TC:? 0  0xe10ffc7 tee_ta_init_user_ta_session at optee_os/core/arch/arm/kernel/user_ta.c:1020
E/TC:? 0  0xe12a7bf tee_ta_init_session at optee_os/core/kernel/tee_ta_manager.c:511
E/TC:? 0  0xe12a83f tee_ta_open_session at optee_os/core/kernel/tee_ta_manager.c:538
E/TC:? 0  0xe120f93 entry_open_session at optee_os/core/arch/arm/tee/entry_std.c:304
E/TC:? 0  0xe102e71 tee_entry_std at optee_os/core/arch/arm/tee/entry_std.c:550
E/TC:? 0  0xe1135e5 __thread_std_smc_entry at optee_os/core/arch/arm/kernel/thread.c:594
E/TC:? 0  0xe101650 thread_std_smc_entry at optee_os/core/arch/arm/kernel/thread_a32.S:363
E/TC:? 0 ...while holding lock 0xe15d110 acquired at:
E/TC:? 0 Call stack:
E/TC:? 0  0xe129131 __lockdep_lock_acquire at optee_os/core/kernel/lockdep.c:309
E/TC:? 0  0xe11653d lockdep_lock_acquire at optee_os/core/include/kernel/lockdep.h:82
E/TC:? 0  0xe1166a9 mutex_lock_check at optee_os/core/arch/arm/kernel/mutex_lockdep.c:45
E/TC:? 0  0xe115e5b __mutex_lock at optee_os/core/arch/arm/kernel/mutex.c:31
E/TC:? 0  0xe116229 mutex_lock at optee_os/core/arch/arm/kernel/mutex.c:238
E/TC:? 0  0xe12a74f tee_ta_init_session at optee_os/core/kernel/tee_ta_manager.c:495
E/TC:? 0  0xe12a83f tee_ta_open_session at optee_os/core/kernel/tee_ta_manager.c:538
E/TC:? 0  0xe120f93 entry_open_session at optee_os/core/arch/arm/tee/entry_std.c:304
E/TC:? 0  0xe102e71 tee_entry_std at optee_os/core/arch/arm/tee/entry_std.c:550
E/TC:? 0  0xe1135e5 __thread_std_smc_entry at optee_os/core/arch/arm/kernel/thread.c:594
E/TC:? 0  0xe101650 thread_std_smc_entry at optee_os/core/arch/arm/kernel/thread_a32.S:363
E/TC:? 0 -> Thread 0xe15da08 acquired lock 0xe15d110 at:
E/TC:? 0 Call stack:
E/TC:? 0  0xe129131 __lockdep_lock_acquire at optee_os/core/kernel/lockdep.c:309
E/TC:? 0  0xe11653d lockdep_lock_acquire at optee_os/core/include/kernel/lockdep.h:82
E/TC:? 0  0xe1166a9 mutex_lock_check at optee_os/core/arch/arm/kernel/mutex_lockdep.c:45
E/TC:? 0  0xe115e5b __mutex_lock at optee_os/core/arch/arm/kernel/mutex.c:31
E/TC:? 0  0xe116229 mutex_lock at optee_os/core/arch/arm/kernel/mutex.c:238
E/TC:? 0  0xe129f77 tee_ta_clear_busy at optee_os/core/kernel/tee_ta_manager.c:145
E/TC:? 0  0xe12aa35 tee_ta_invoke_command at optee_os/core/kernel/tee_ta_manager.c:627
E/TC:? 0  0xe12114d entry_invoke_command at optee_os/core/arch/arm/tee/entry_std.c:375
E/TC:? 0  0xe102e8d tee_entry_std at optee_os/core/arch/arm/tee/entry_std.c:556
E/TC:? 0  0xe1135e5 __thread_std_smc_entry at optee_os/core/arch/arm/kernel/thread.c:594
E/TC:? 0  0xe101650 thread_std_smc_entry at optee_os/core/arch/arm/kernel/thread_a32.S:363
E/TC:? 0  0xe178718 __heap1_start at :?
E/TC:? 0 ...while holding lock 0xe15d144 acquired at:
E/TC:? 0 Call stack:
E/TC:? 0  0xe129131 __lockdep_lock_acquire at optee_os/core/kernel/lockdep.c:309
E/TC:? 0  0xe11653d lockdep_lock_acquire at optee_os/core/include/kernel/lockdep.h:82
E/TC:? 0  0xe1166a9 mutex_lock_check at optee_os/core/arch/arm/kernel/mutex_lockdep.c:45
E/TC:? 0  0xe115e5b __mutex_lock at optee_os/core/arch/arm/kernel/mutex.c:31
E/TC:? 0  0xe116229 mutex_lock at optee_os/core/arch/arm/kernel/mutex.c:238
E/TC:? 0  0xe136229 ree_fs_closedir_rpc at optee_os/core/tee/tee_ree_fs.c:957
E/TC:? 0  0xe1339af tee_svc_close_enum at optee_os/core/tee/tee_svc_storage.c:89
E/TC:? 0  0xe134483 syscall_storage_free_enum at optee_os/core/tee/tee_svc_storage.c:629
E/TC:? 0  0xe10268c tee_svc_do_call at optee_os/core/arch/arm/tee/arch_svc_a32.S:60
E/TC:? 0  0xe10ef37 user_ta_enter at optee_os/core/arch/arm/kernel/user_ta.c:320
E/TC:? 0  0xe10efd9 user_ta_enter_invoke_cmd at optee_os/core/arch/arm/kernel/user_ta.c:350
E/TC:? 0  0xe12aa0d tee_ta_invoke_command at optee_os/core/kernel/tee_ta_manager.c:614
E/TC:? 0  0xe12114d entry_invoke_command at optee_os/core/arch/arm/tee/entry_std.c:375
E/TC:? 0  0xe102e8d tee_entry_std at optee_os/core/arch/arm/tee/entry_std.c:556
E/TC:? 0 lockdep_lock_acquire:85 lockdep: error 0xffff0007
E/TC:0 0 Panic at core/include/kernel/lockdep.h:86 <lockdep_lock_acquire>
E/TC:0 0 Call stack:
E/TC:0 0  0x0e11895b print_kernel_stack at optee_os/core/arch/arm/kernel/unwind_arm32.c:464
E/TC:0 0  0x0e1299dd __do_panic at optee_os/core/kernel/panic.c:32 (discriminator 1)
E/TC:0 0  0x0e116587 lockdep_lock_acquire at optee_os/core/include/kernel/lockdep.h:88
E/TC:0 0  0x0e1166a9 mutex_lock_check at optee_os/core/arch/arm/kernel/mutex_lockdep.c:45
E/TC:0 0  0x0e115e5b __mutex_lock at optee_os/core/arch/arm/kernel/mutex.c:31
E/TC:0 0  0x0e116229 mutex_lock at optee_os/core/arch/arm/kernel/mutex.c:238
E/TC:0 0  0x0e129f77 tee_ta_clear_busy at optee_os/core/kernel/tee_ta_manager.c:145
E/TC:0 0  0x0e12aa35 tee_ta_invoke_command at optee_os/core/kernel/tee_ta_manager.c:627
E/TC:0 0  0x0e12114d entry_invoke_command at optee_os/core/arch/arm/tee/entry_std.c:375
E/TC:0 0  0x0e102e8d tee_entry_std at optee_os/core/arch/arm/tee/entry_std.c:556
E/TC:0 0  0x0e1135e5 __thread_std_smc_entry at optee_os/core/arch/arm/kernel/thread.c:594
E/TC:0 0  0x0e101650 thread_std_smc_entry at optee_os/core/arch/arm/kernel/thread_a32.S:363
```

File: core/arch/arm/include/kernel/unwind.h [hunk 1]
```diff
@@ -113,13 +113,19 @@ static inline void print_kernel_stack(int level __unused)
 
 #ifdef CFG_UNWIND
 TEE_Result relocate_exidx(void *exidx, size_t exidx_sz, int32_t offset);
+/* Get current call stack as an array allocated on the heap */
+vaddr_t *get_kernel_stack(void);
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/include/kernel/unwind.h]: True. I can call it `unw_get_kernel_stack()` if that's better. Then we may rename the other functions in a separate patch.

File: core/arch/arm/kernel/mutex_lockdep.c [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+	for (n = 0; n < CFG_NUM_THREADS; n++)
+		TAILQ_INIT(&owned[n]);
+
+	DMSG("lockdep is enabled for mutexes");
+}
+
+void mutex_lock_check(struct mutex *m)
+{
+	int thread = thread_get_id_may_fail();
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/kernel/mutex_lockdep.c]: My mistake, will use `thread_get_id()`.

File: core/kernel/lockdep.c [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+				TAILQ_INSERT_TAIL(&queue, nqe, link);
+			}
+		}
+		free(qe->path);
+		free(qe);
+		qe = NULL;
+	}
+
+out:
+	free(qe);
```
Comment (jbech-linaro) [inline core/kernel/lockdep.c]: `free(qe->path);` also, we have several `goto out` in other places after a successful malloc.

File: core/kernel/lockdep.c [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+		}
+	}
+	free(cycle);
+}
+
+TEE_Result __lockdep_lock_acquire(struct lockdep_node_head *graph,
+				  struct lockdep_lock_head *owned,
+				  uintptr_t id)
+{
+	vaddr_t *acq_stack = get_kernel_stack();
```
Comment (jbech-linaro) [inline core/kernel/lockdep.c]: This could be moved after the `if (!node)` check.

File: core/kernel/lockdep.c (hunk 3 above)
Comment (jbech-linaro) [reply core/kernel/lockdep.c]: Ah, wait a bit `ret = qe->path;` ... kind of hides that. So my previous comment here might be an incorrect claim.

File: core/arch/arm/kernel/mutex_lockdep.c [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+#include <sys/queue.h>
+#include <trace.h>
+
+#include "mutex_lockdep.h"
+
+/* Global graph of all mutexes used in the code */
+static struct lockdep_node_head graph = TAILQ_HEAD_INITIALIZER(graph);
+
+/* Protects @graph */
+static struct mutex m_graph = MUTEX_INITIALIZER;
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/mutex_lockdep.c]: How about using a spinlock instead to protect the lockdep state?

File: core/arch/arm/kernel/unwind_arm32.c [hunk 6]
```diff
@@ -483,3 +482,63 @@ void print_kernel_stack(int level)
 }
 
 #endif
+
+#if defined(CFG_UNWIND) && defined(ARM32)
+static void *grow(void *p, size_t *cur_size, size_t new_size)
+{
+	size_t rounded_size = ROUNDUP(new_size, 16 * sizeof(vaddr_t));
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/kernel/unwind_arm32.c]: Right, will change

File: core/arch/arm/kernel/unwind_arm32.c [hunk 7]
```diff
@@ -483,3 +482,63 @@ void print_kernel_stack(int level)
...
+	void *tmp = NULL;
+
+	if (*cur_size >= new_size)
+		return p;
+	tmp = realloc(p, rounded_size);
+	if (tmp)
+		*cur_size = rounded_size;
+	return tmp;
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/kernel/unwind_arm32.c]: If the allocation fails, `unw_get_kernel_stack()` will return NULL and there will be no stack recorded/printed. I can add an `EMSG()` in the `err:` label of `unw_get_kernel_stack()` if you like.

File: core/arch/arm/kernel/unwind_arm64.c [hunk 8]
```diff
@@ -96,3 +95,56 @@ void print_kernel_stack(int level)
 }
 
 #endif
+
+#if defined(CFG_UNWIND)
+static void *grow(void *p, size_t *cur_size, size_t new_size)
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/kernel/unwind_arm64.c]: Right, how about I move it to `unwind_private.h` (new file) as a `static inline` func?

File: core/include/kernel/lockdep.h [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+#include <sys/queue.h>
+#include <tee_api_types.h>
+#include <trace.h>
+#include <types_ext.h>
+
+/*
+ * Lock graph. If node A has an edge to node B, then A was locked before B in
+ * the same thread of execution.
+ */
+
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/include/kernel/lockdep.h]: Well, the comment does not apply only to the `struct lockdep_edge`, you need the two structs and the two `STAILQ_HEAD()` to define a graph.

File: core/kernel/lockdep.c [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+#define LOCKDEP_NODE_BFS_VISITED	BIT(2)
+
+/* Find node in graph or add it */
+static struct lockdep_node *lockdep_add_to_graph(
+				struct lockdep_node_head *graph,
+				uintptr_t lock_id)
+{
+	struct lockdep_node *node = NULL;
+
+	TAILQ_FOREACH(node, graph, link)
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/kernel/lockdep.c]: That would be a programming error, so I can add an `assert()` here.

File: core/arch/arm/kernel/mutex_lockdep.c (hunk 5 above)
Comment (jforissier) [reply core/arch/arm/kernel/mutex_lockdep.c]: Good idea, will do.

File: core/arch/arm/kernel/unwind_arm32.c [hunk 11]
```diff
@@ -483,3 +482,63 @@ void print_kernel_stack(int level)
...
+
+vaddr_t *get_kernel_stack(void)
+{
+	size_t n = 0;
+	size_t size = 0;
+	vaddr_t *tmp = NULL;
+	vaddr_t *addr = NULL;
+	struct unwind_state_arm32 state;
+	uaddr_t exidx = (vaddr_t)__exidx_start;
+	size_t exidx_sz = (vaddr_t)__exidx_end - (vaddr_t)__exidx_start;
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/kernel/unwind_arm32.c]: OK

**Comment by jforissier** [comment]
Addressed review comments and checkpatch warnings (remaining ones are caused by the BSD queue macros, I think we should just ignore them).

File: core/kernel/lockdep.c [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+	nstack = malloc(n * sizeof(vaddr_t));
+	if (!nstack)
+		return NULL;
+	memcpy(nstack, stack, n * sizeof(vaddr_t));
+	return nstack;
+}
+
+static void lockdep_print_call_stack(vaddr_t *stack)
+{
+	vaddr_t *p = NULL;
```
Comment (etienne-lms) [inline core/kernel/lockdep.c]: min,or: useless init

File: core/kernel/lockdep.c [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+		EMSG_RAW(" %#" PRIxPTR, *p);
+}
+
+static TEE_Result lockdep_add_edge(struct lockdep_node *from,
+				   struct lockdep_node *to,
+				   vaddr_t *call_stack_from,
+				   vaddr_t *call_stack_to,
+				   uintptr_t thread_id)
+{
+	struct lockdep_edge *edge = NULL;
```
Comment (etienne-lms) [inline core/kernel/lockdep.c]: minor: useless init

File: core/kernel/lockdep.c [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+	int pathlen;
+	TAILQ_ENTRY(lockdep_bfs) link;
+};
+
+TAILQ_HEAD(lockdep_bfs_head, lockdep_bfs);
+
+static void lockdep_bfs_queue_delete(struct lockdep_bfs_head *queue)
+{
+	struct lockdep_bfs *cur = NULL;
+	struct lockdep_bfs *next = NULL;
```
Comment (etienne-lms) [inline core/kernel/lockdep.c]: minor: inits required?

File: core/kernel/lockdep.c [hunk 15]
```diff
@@ -0,0 +1,397 @@
...
+/*
+ * Print shortest cycle in @graph that contains @node.
+ * This function performs an iterative breadth-first search starting from @node,
+ * and stops when it reaches @node again. In each node we're tracking the path
+ * from the start node.
+ */
+static uintptr_t *lockdep_graph_get_shortest_cycle(struct lockdep_node *node)
+{
+	struct lockdep_bfs_head queue;
+	struct lockdep_bfs *qe = NULL;
```
Comment (etienne-lms) [inline core/kernel/lockdep.c]: minor: useless init

File: core/kernel/lockdep.c (hunk 3 above)
Comment (etienne-lms) [reply core/kernel/lockdep.c]: If allocation fails at line 157, the caller may print out incomplete information without noticing the error.
Should this function return NULL in this case and locally free the allocated `qe->path`?

File: core/kernel/lockdep.c [hunk 16]
```diff
@@ -0,0 +1,397 @@
...
+	if (!qe->path)
+		goto out;
+	qe->path[0] = node->lock_id;
+	qe->pathlen = 1;
+	TAILQ_INSERT_TAIL(&queue, qe, link);
+
+	while (!TAILQ_EMPTY(&queue)) {
+		qe = TAILQ_FIRST(&queue);
+
+		struct lockdep_node *n = qe->node;
```
Comment (etienne-lms) [inline core/kernel/lockdep.c]: minor: unused, remove.

File: core/kernel/lockdep.c [hunk 17]
```diff
@@ -0,0 +1,397 @@
...
+				goto out;
+			}
+
+			if (!(e->to->flags & LOCKDEP_NODE_BFS_VISITED)) {
+				e->to->flags |= LOCKDEP_NODE_BFS_VISITED;
+
+				size_t nlen = qe->pathlen + 1;
+				struct lockdep_bfs *nqe = calloc(1,
+								 sizeof(*nqe));
+				if (!nqe)
```
Comment (etienne-lms) [inline core/kernel/lockdep.c]: minor: add an empty line

File: core/kernel/lockdep.c [hunk 18]
```diff
@@ -0,0 +1,397 @@
...
+			return res;
+	}
+
+	node->flags |= LOCKDEP_NODE_PERM_MARK;
+	return TEE_SUCCESS;
+}
+
+static TEE_Result lockdep_graph_sort(struct lockdep_node_head *graph)
+{
+	struct lockdep_node *node = NULL;
```
Comment (etienne-lms) [inline core/kernel/lockdep.c]: minor: init required?

File: core/kernel/lockdep.c [hunk 19]
```diff
@@ -0,0 +1,397 @@
...
+		node->flags = 0;
+
+	return TEE_SUCCESS;
+}
+
+static struct lockdep_edge *lockdep_find_edge(struct lockdep_node_head *graph,
+					      uintptr_t from, uintptr_t to)
+{
+	struct lockdep_node *node = NULL;
+	struct lockdep_edge *edge = NULL;
```
Comment (etienne-lms) [inline core/kernel/lockdep.c]: inits?

File: core/kernel/lockdep.c [hunk 20]
```diff
@@ -0,0 +1,397 @@
...
+				  struct lockdep_lock_head *owned,
+				  uintptr_t id)
+{
+	vaddr_t *acq_stack = get_kernel_stack();
+	struct lockdep_node *node = lockdep_add_to_graph(graph, id);
+
+	if (!node)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	struct lockdep_lock *lock = NULL;
```
Comment (etienne-lms) [inline core/kernel/lockdep.c]: init

File: core/arch/arm/kernel/mutex_lockdep.c [hunk 21]
```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018, Linaro Limited
+ */
+
+#include <assert.h>
+#include <kernel/thread.h>
```
Comment (etienne-lms) [inline core/arch/arm/kernel/mutex_lockdep.c]: minor: shift down 1 line

File: mk/config.mk [hunk 22]
```diff
@@ -219,9 +219,17 @@ CFG_TA_DYNLINK ?= y
 # Enable paging, requires SRAM, can't be enabled by default
 CFG_WITH_PAGER ?= n
 
+# Runtime lock dependency checker: ensures that a proper locking hierarchy is
+# used in the TEE core when acquiring and releasing mutexes. Any violation will
+# cause a panic as soon as the invalid locking condition is detected. If
+# CFG_UNWIND is enabled, the algorithm records the call stacks when locks are
+# taken, and prints them when a potential deadlock is found.
+# Expect a significant performance impact when enbling this.
```
Comment (etienne-lms) [inline mk/config.mk]: s/enbling/enabling/

File: core/arch/arm/kernel/unwind_private.h [hunk 23]
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018, Linaro Limited
+ */
+#include <malloc.h>
+#include <stddef.h>
+#include <util.h>
+
+#ifdef CFG_UNWIND
```
Comment (etienne-lms) [inline core/arch/arm/kernel/unwind_private.h]: minor: `#ifdef CFG_UNWIND` could be removed.

File: core/arch/arm/kernel/unwind_private.h [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+static inline void *unw_grow(void *p, size_t *cur_size, size_t new_size)
+{
+	if (*cur_size >= new_size)
+		return p;
+
+	size_t rounded_size = ROUNDUP(new_size, 16 * sizeof(vaddr_t));
+
+	void *tmp = realloc(p, rounded_size);
```
Comment (etienne-lms) [inline core/arch/arm/kernel/unwind_private.h]: minor: add en empty line. maybe remove the one between the 2 local variables definition.

File: core/arch/arm/kernel/mutex_lockdep.c [hunk 25]
```diff
@@ -36,23 +37,19 @@ void mutex_lockdep_init(void)
 void mutex_lock_check(struct mutex *m)
 {
 	int thread = thread_get_id();
+	uint32_t exceptions = 0;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/mutex_lockdep.c]: minor: useless init.

File: core/arch/arm/pta/core_self_tests.h [hunk 26]
```diff
@@ -18,4 +19,17 @@ TEE_Result core_fs_htree_tests(uint32_t nParamTypes,
...
+TEE_Result core_lockdep_tests(uint32_t nParamTypes,
+			      TEE_Param pParams[TEE_NUM_PARAMS]);
+#else
+static inline TEE_Result core_lockdep_tests(
+		uint32_t nParamTypes __unused,
+		TEE_Param pParams[TEE_NUM_PARAMS] __unused)
+{
+	return TEE_ERROR_NOT_SUPPORTED;
+}
+
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/pta/core_self_tests.h]: ok

File: core/kernel/lockdep.c (hunk 12 above)
Comment (jforissier) [reply core/kernel/lockdep.c]: Yeah, it's a new rule, for security reasons we're now requiring that all local variables be initialized.
https://github.com/OP-TEE/optee_os/commit/439203cb04ce

File: core/arch/arm/kernel/mutex_lockdep.c (hunk 21 above)
Comment (jforissier) [reply core/arch/arm/kernel/mutex_lockdep.c]: Has been changed already

File: mk/config.mk (hunk 22 above)
Comment (jforissier) [reply mk/config.mk]: Will fix

File: core/arch/arm/kernel/unwind_private.h (hunk 23 above)
Comment (jforissier) [reply core/arch/arm/kernel/unwind_private.h]: True, will do

File: core/arch/arm/kernel/unwind_private.h (hunk 24 above)
Comment (jforissier) [reply core/arch/arm/kernel/unwind_private.h]: Yep

**Comment by jforissier** [comment]
Addressed Etienne's comments (thanks!).

File: core/kernel/lockdep.c (hunk 12 above)
Comment (etienne-lms) [reply core/kernel/lockdep.c]: ok, thanks. I've removed my related comments.

File: core/arch/arm/pta/core_lockdep_tests.c [hunk 27]
```diff
@@ -0,0 +1,149 @@
...
+	int count = 0;
+
+	DMSG("");
+
+	TAILQ_INIT(&thread1);
+	TAILQ_INIT(&graph);
+
+	/* Not locked, expect failure */
+	res = __lockdep_lock_release(&thread1, 1);
+	if (!res)
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/pta/core_lockdep_tests.c]: Good catch! Will start counting at 1.

File: core/arch/arm/kernel/thread.c [hunk 28]
```diff
@@ -12,6 +12,7 @@
 #include <kernel/asan.h>
 #include <kernel/misc.h>
 #include <kernel/msg_param.h>
+#include <kernel/lockdep.h>
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread.c]: minor: shift up

File: core/arch/arm/kernel/unwind_arm64.c [hunk 29]
```diff
@@ -96,3 +97,43 @@ void print_kernel_stack(int level)
 }
 
 #endif
+
+#if defined(CFG_UNWIND)
```
Comment (etienne-lms) [inline core/arch/arm/kernel/unwind_arm64.c]: Useless `#ifdef CFG_UNWIND`. The same comment applies to change in **unwind_a32.c**. 
Feel free to address or not this comment...

File: core/arch/arm/pta/core_lockdep_tests.c [hunk 30]
```diff
@@ -0,0 +1,149 @@
...
+		return count;
+
+	lockdep_graph_delete(&graph);
+	lockdep_queue_delete(&thread1);
+
+	return 0;
+}
+
+static int __maybe_unused self_test_lockdep2(void)
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/pta/core_lockdep_tests.c]: joakim's comment still applies... same below.

File: core/arch/arm/pta/core_lockdep_tests.c [hunk 31]
```diff
@@ -0,0 +1,149 @@
...
+	res = __lockdep_lock_acquire(&graph, &thread2, 3);
+	if (res)
+		return count;
+	count++;
+
+	/* Deadlock 1-2-3 */
+	res = __lockdep_lock_acquire(&graph, &thread3, 1);
+	if (!res)
+		return count;
+	count++;
```
Comment (etienne-lms) [inline core/arch/arm/pta/core_lockdep_tests.c]: minor: can remove this increment. ditto in function below.

File: core/arch/arm/pta/core_self_tests.h [hunk 32]
```diff
@@ -5,6 +5,7 @@
 #ifndef CORE_SELF_TESTS_H
 #define CORE_SELF_TESTS_H
 
+#include <compiler.h>
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/pta/core_self_tests.h]: i guess for the `__unused ` attribute.

File: core/kernel/lockdep.c (hunk 16 above)
Comment (etienne-lms) [reply core/kernel/lockdep.c]: `n` unused.

File: core/arch/arm/kernel/thread.c (hunk 28 above)
Comment (jforissier) [reply core/arch/arm/kernel/thread.c]: ok

File: core/arch/arm/kernel/unwind_arm64.c (hunk 29 above)
Comment (jforissier) [reply core/arch/arm/kernel/unwind_arm64.c]: OK, and I will also clean up the others in a separate commit.

File: core/arch/arm/pta/core_lockdep_tests.c [hunk 33]
```diff
@@ -0,0 +1,149 @@
...
+ * Test lockdep with hypothetical thread and lock objects
+ */
+
+#include <assert.h>
+#include <compiler.h>
+#include <kernel/lockdep.h>
+
+#include "core_self_tests.h"
+
+static int __maybe_unused self_test_lockdep1(void)
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/pta/core_lockdep_tests.c]: Right, leftover from early tests, will remove.

File: core/kernel/lockdep.c (hunk 16 above)
Comment (etienne-lms) [reply core/kernel/lockdep.c]: looks like you really don't want to get rid of this `n` :)

File: core/kernel/lockdep.c (hunk 17 above)
Comment (etienne-lms) [reply core/kernel/lockdep.c]: missing empty line above.

File: core/kernel/lockdep.c (hunk 16 above)
Comment (jforissier) [reply core/kernel/lockdep.c]: It is used on line 151, actually ^_^

Comment (etienne-lms) [reply core/kernel/lockdep.c]: oh.. yes. sorry.

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by jforissier** [comment]
@jenswi-linaro since you commented on one patch, do you want to give some Ack/Review tag?

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Pull request merged**